### ISSUE TITLE
errors with custom compiler folder path

### ISSUE THREAD

**Issue by jcodewalker** [vdm-io/Joomla-Component-Builder#571]
### Steps to reproduce the issue
add a custom folder path to "Compiler folder path" field in JCB component configuration.

### Expected result
demo component correctly compiled .

### Actual result
long list of errors (obviously all files listed as missing are correctly in that folder as I moved all the administrator/components/com_componentbuilder/compiler folder content there following the instructions)
and
demo component half compiled (see attached images)
![empty-controllers-folder](https://user-images.githubusercontent.com/49115855/84587226-b314f080-ae1d-11ea-971f-e8a3193b6010.png)
![errors-list](https://user-images.githubusercontent.com/49115855/84587228-b4461d80-ae1d-11ea-8fea-033764e7488f.png)
![helpers-folder](https://user-images.githubusercontent.com/49115855/84587229-b5774a80-ae1d-11ea-9e1a-973e8aaacb26.png)
![layout-folder](https://user-images.githubusercontent.com/49115855/84587230-b60fe100-ae1d-11ea-9f32-8a209691ca94.png)
![models-folder](https://user-images.githubusercontent.com/49115855/84587231-b6a87780-ae1d-11ea-9bb2-e8244d2a52d3.png)
![views-folder](https://user-images.githubusercontent.com/49115855/84587234-b7d9a480-ae1d-11ea-8aa2-498ef9e12ada.png)

### System information (as much as possible)
- OS Name & Version: 
- MySql Version: 
- Apache Version: 
- PHP Version: 
- Joomla Version: both 3.9.19 and 3.9.11
- JCB Version: both 2.10.13 and 2.11.2
- Browser: 

### Additional comments
Compiling a more complex component with plugins and modules, I see that plugins and modules are correctly compiled, while the component is still half compiled like demo component.

**Comment by Llewellynvdm** [comment]
Make sure that your web-server has access to that file location.

May I ask why you moved the compiler folder? When updating JCB the installer (of Joomla) will not move the new files (of the compiler) to update the remote folder, heads-up on that one.

Anyway are you on Mac or Windows? The file system is a little tricky to say the least, I would advice that if you are having issue just move stuff back to the default paths and all should be okay.

The when adding custom folders you will need to manually update those as you update JCB. Very important.

Then you must use full paths.

Remember the Joomla's own File class is used to move files around, so JCB is not really doing the work, when you see errors, those errors (file or folder errors) actually come from Joomla. So if the Joomla File&Folder classes can't reach those files there very little I can do to fix that.

The plugin, and modules are build in another way, and using no template files, where the components are still heavily dependent on them, I am slowly busy changing that.

**Comment by jcodewalker** [comment]
I’m on Mac.

I already moved outside the custom folder and it works perfectly.

In the same directory I moved the compiler folder and its content and obviously the folder has the correct access rights and can be accessed because both site and administrator component’s main file (“componentname.php”, for the demo component it’s “demo.php”) are correctly created using /compiler/Joomla_3/component_admin.php and /compiler/Joomla_3/component_site.php. Yes, JCB really uses correctly those files because if I edit these files the compiled files contains those modifications. So, there are no doubs that JCB can access exactly those files.

So, it’s not an environment issue, it’s not a Joomla issue and yes, it’s a JCB bug.

It can be something similar to the module’s custom fields… a wrong path declaration when JCB compiles only the files inside the folders, so it can be related to folders path declaration.

Hope it helps.

BTW, I needed to change the structure to fix my needs, I moved the item find check (and redirect) from model to view, edited the dashboard page and other things. So yes, I need to avoid to update those files. It would be interesting to move only the templates folder (Joomla_3), which I think changes only if Joomla structure changess (i.e. Joomla_4).

And as I have more than 1 JCB installation, moving something outside the installation folders let me share some common things with the different JCB installations I have. As you can see each time we update JCB there are bugs here and there. In this way, I can update just one installation, check it out, and if there are bugs that are limiting the production… I still have others running perfectly. It’s not the best, I know, a lot of work duplicated… but we have to survive to the bugs… :-)

**Comment by Llewellynvdm** [comment]
This is an excellent work flow :+1: 

About the file path declarations... this is a little more challenging. Meaning I can't see that the issue is there, I think they are correct. Here is why, all paths are setup only once and each in only one place, which makes it very easy to check the file paths. (I will share here where those moments or areas are found, and you can check in your setup with the `var_dump($this->path_name_here);jexit();` method what are in those values, and if they are indeed correct.

This [line set up the compiler](https://github.com/vdm-io/Joomla-Component-Builder/blob/staging/admin/helpers/compiler/a_Get.php#L871) path:
```php
// load the compiler path
$this->compilerPath = $this->params->get(
	'compiler_folder_path',
	JPATH_COMPONENT_ADMINISTRATOR . '/compiler'
);
```
This [line set up the template and custom folder](https://github.com/vdm-io/Joomla-Component-Builder/blob/staging/admin/helpers/compiler/b_Structure.php#L395) paths:
```php
// set the template path
$this->templatePath = $this->compilerPath . '/joomla_'
	. $config['version'];
// set some default names
$this->componentSalesName  = 'com_'
	. $this->componentData->sales_name . '__J'
	. $this->joomlaVersion;
$this->componentBackupName = 'com_'
	. $this->componentData->sales_name . '_v' . str_replace(
		'.', '_', $this->componentData->component_version
	) . '__J' . $this->joomlaVersion;
$this->componentFolderName = 'com_'
	. $this->componentData->name_code . '_v' . str_replace(
		'.', '_', $this->componentData->component_version
	) . '__J' . $this->joomlaVersion;
// set component folder path
$this->componentPath = $this->compilerPath . '/'
	. $this->componentFolderName;
// set the template path for custom
$this->templatePathCustom = $this->params->get(
	'custom_folder_path', JPATH_COMPONENT_ADMINISTRATOR . '/custom'
);
```
These paths are never build again, and so we reuse it all over the build process. So as you can see it just takes the values.

So taking for example the `$this->templatePath` value doing a search in the [same b_Structure.php file](https://github.com/vdm-io/Joomla-Component-Builder/blob/staging/admin/helpers/compiler/b_Structure.php) you will see where ever it is used, do a few check at these various places to see it something changed.

I tried to keep all the structural things in the b_Structure.php file, but there are some functions needed earlier in the build so they are added to the [a_Get.php](https://github.com/vdm-io/Joomla-Component-Builder/blob/staging/admin/helpers/compiler/a_Get.php) file. So between these two files most of the file structural functions and work is done, even for the plugin, and module paths.

[Here you see the module paths](https://github.com/vdm-io/Joomla-Component-Builder/blob/staging/admin/helpers/compiler/b_Structure.php#L498) are set based on above base paths:
``php
$module->folder_path = $this->compilerPath . '/'
	. $module->folder_name;
```

[Here you see the plugin paths](https://github.com/vdm-io/Joomla-Component-Builder/blob/staging/admin/helpers/compiler/b_Structure.php#L1030) are set based on above base paths:
``php
$plugin->folder_path = $this->compilerPath . '/'
	. $plugin->folder_name;
```

Now the actual module folder name is set in the [a_Get.php on line 8213](https://github.com/vdm-io/Joomla-Component-Builder/blob/staging/admin/helpers/compiler/a_Get.php#L8213) and the actual plugin folder name is also set int the [a_Get.php on line 9135](https://github.com/vdm-io/Joomla-Component-Builder/blob/staging/admin/helpers/compiler/a_Get.php#L9135).

Then to be able to improve JCB, you should, you are welcome to, open issues with the recommended file changes where you think there are better options (speaking about the templates mostly here).

**Comment by Llewellynvdm** [comment]
I found the issue, Joomla made a change to the copy method and JCB was still using the old implementation in two places. Please test the staging branch, as this issue should now be resolved.

**Comment by jcodewalker** [comment]
Simply amazing!

Yes, it works fine. I’ve just tested with the demo component in a brand new installation and it was compiled.

GREAT!
Thank you.

**Comment by jcodewalker** [comment]
Hello,

About the last part ("Then to be able to improve JCB, you should, you are welcome to, open issues with the recommended file changes where you think there are better options (speaking about the templates mostly here).”), I don’t think it is in terms of better options, for this I didn’t open a new issue, but more in a personal work flow terms, which I’m going to share with you (if you are interested).

Generally speaking, the model is the place were we are supposed to manage data, and only this. In the site side, item model (single item) I avoid to redirect the user to the main view (or to the base root if there isn’t any main view, e_interpretation.php row #3563 <https://github.com/vdm-io/Joomla-Component-Builder/blob/staging/admin/helpers/compiler/e_Interpretation.php#L3563> ) directly in the model. I’ve changed this and in the if statement I return false instead. Then it is in the view that I (as developer) can decide what I want to do with it. I can trigger a plugin and then redirect the user where I want (maybe I set a custom redirect link in the component I want to use, or something else). It started for me because the JCB redirect doesn’t work in my installations, not local nor live. I always get an error saying that the page redirects in a wrong way. You know, you can't know if a component looks for $menu->params to refine the dynamic get and in that case a menu id (“&Itemid=xxx”) is required in the url.
Long story short, in the getItem method, if $data is empty the model returns false (so the model manages data, what is supposed to do) and in the view I can decide what to do in case of $this->item === false. Personally, I prefer to show a “404 layout” called in views/myview/default.php file (so the customer can even override it) with a custom message, after trigger a plugin in the views/myview/view.html.php file. So, you see, I’m not sure it is an improvement but surely it fixed an issue I found with my components.

Then, still in getItem method, single item model, site side, if the dynamic get has a Join Inner joint, in the model we have a sub method called like $this->getId{Maintable}{Joinedtable}_R($data->id); . Even if it is called within a try-catch loop, actually it doesn’t throw an error if it fails (I went mad one day to find out what was wrong!). So, I replaced every db call with something different, a class which take care of the db result and take trace of errors, the method and the line where the db call started (because I pass it from the e_interpretation.php with __METHOD__.’ line: ‘.__LINE__ string), the db error returned and the query string, saving all in a file located in administrator/logs/com_componentname.php, before to return the db result (false in case of errors). Maybe this can be an improvement? I don’t know honestly.

So, as you can see aren’t that better options, but they help me a lot to have a better control of what is going on in the extensions I develop.

Hope it helps.
Let me know if you need more info.

Best regards,
claudio
